pkg/sql/catalog/tabledesc/validate.go

```diff
@@ -640,6 +640,40 @@ func (desc *wrapper) ValidateSelf(vea catalog.ValidationErrorAccumulator) {
 		vea.Report(err)
 		return nil
 	})
+
+	// Validate that there are no column with both a foreign key ON UPDATE and an
+	// ON UPDATE expression. This check is made to ensure that we know which ON
+	// UPDATE action to perform when a FK UPDATE happens.
+	if err := ValidateOnUpdate(desc.AllColumns(), desc.GetOutboundFKs()); err != nil {
+		vea.Report(err)
+	}
+}
+
+// ValidateOnUpdate returns an error if there is a column with both a foreign
+// key constraint and an ON UPDATE expression, nil otherwise.
+func ValidateOnUpdate(cols []catalog.Column, outboundFks []descpb.ForeignKeyConstraint) error {
+	for _, col := range cols {
+		if !col.HasOnUpdate() {
+			continue
+		}
+		for _, fk := range outboundFks {
+			if fk.OnUpdate == descpb.ForeignKeyReference_NO_ACTION ||
+				fk.OnUpdate == descpb.ForeignKeyReference_RESTRICT {
+				continue
+			}
+			for _, fkCol := range fk.OriginColumnIDs {
+				if fkCol == col.GetID() {
+					return errors.AssertionFailedf(
+						"cannot specify both ON UPDATE expression and a foreign key"+
+							" ON UPDATE action for column %s(%d)",
+						col.GetName(),
+						col.GetID(),
```

Comment:
This error also doesn't need to be an assertion failure, you can return a `pgerror.New(pgcode.InvalidTableDefinition, ...`
The validation framework will wrap these into assertion failures when required (at descriptor read and write time).

Also, ideally we try to capture as many validation failures as possible, not just the first encountered. I understand you want to use this in `create_table.go` as well. Perhaps there's a way. If you look at other validation code you'll notice that we're not particularly disciplined in doing this but better late than never.